Add test suite

ionthegeek opened the pull request:
# Summary
This PR adds the beginnings of a test suite to `promdump` by adding tests for the functions `getMetricName` and `getRangeTimestamps` since these are the lowest hanging fruit.

## getMetricName Tests
* mutual exclusivity of exportName and jobName must be enforced
* getMetricName must return the export name for exporter metrics
* getMetricName must return the job name for job metrics

## getRangeTimestamps Tests
* at most two of startTime, endTime, and period can be specified
* endTime should default to now if neither the start time nor end time are specified
* endTime should default to now if only the period is specified
* when both start time and end time are specified, the period is the difference between the two
* when none of start time, end time, or period are specified, the period should be the default period
* when we have the period and one time stamp, the other time stamp should be time stamp +/- the period
* startTime must be in the past
* startTime must be before endTime
* end timestamp should be set to now if it's a time from the future

Implementing tests for behaviour that depends on the current time required some changes to the core promdump code. All calls to time.Now() were replaced with a func called now() that returns time.Now().

In the test suite, there is a func called freezeTime() that gets the current time from time.Now() and overrides the now() func so that it always returns this static time. This ensures that tests that rely on now() are deterministic.

Commit:
promdump: add a test suite

This commit adds the first test in the test suite. This white box test tests the behaviour of the getMetricName function, verifying that:

1. An error is returned if neither of exportName and jobName are specified for a metric.
2. An error is returned if both of exportName and jobName are specified for a metric.
3. The exportName or jobName is returned if exactly one of those fields is specified.

Commit:
promdump: fix crash when functions under test try to log

Because the functions under test are run in isolation, logging is not set up automatically when they are executed.

Added an init func that initializes the logger to discard log messages generated by the functions under test.

Commit:
promdump: improve solution for setting up logging before tests

Moving logging setup to a setupLogging() func that's called at the beginning of each Test...() func to avoid the potential for weirdness related to having multiple init() funcs in the main package (one for the program code and one for the tests).

Commit:
promdump: add getRangeTimestamps tests to the test suite

This update to the test suite adds test cases for the following:
* at most two of startTime, endTime, and period can be specified
* endTime should default to now if neither the start time nor end time are specified
* endTime should default to now if only the period is specified
* when both start time and end time are specified, the period is the difference between the two
* when none of start time, end time, or period are specified, the period should be the default period
* when we have the period and one time stamp, the other time stamp should be time stamp +/- the period
* startTime must be in the past
* startTime must be before endTime
* end timestamp should be set to now if it's a time from the future

Implementing tests for behaviour that depends on the current time required some changes to the core promdump. All calls to time.Now() were replaced with a func called now() that returns time.Now().

In the test suite, there is a func called freezeTime() that gets the current time from time.Now() and overrides the now() func so that it always returns this static time. This ensures that tests that rely on now() are deterministic.

Commit:
promdump: discard metricName in tests where it is not used

There were two tests where metricName was returned but not used since we were testing for error results. Changed these tests to discard metricName.

na6vj commented on promdump/promdump_test.go:
```diff
@@ -0,0 +1,171 @@
+		"badmetric":   {exportName: "badmetric_export", jobName: "badmetric", collect: true, changedFromDefault: false, requiresNodePrefix: true},
+	}
+
+	testMetric := testMetrics["emptymetric"]
```
This code block is essentially repeated 4 times.

I would suggest putting the err message as a part each metric.
This will allow adding tests without creating more (repeated) code blocks.

na6vj commented on promdump/promdump_test.go:
```diff
@@ -0,0 +1,171 @@
+	// at most two of startTime, endTime, and period can be specified
+	ts := testTimes["tooManyTimes"]
+	_, _, _, err := getRangeTimestamps(ts.startTime, ts.endTime, ts.period)
+	if err == nil {
```
Again, I would suggest a declarative approach - rather than very similar repeated code for each item.

You could put the expected result as a part of the declaration.

na6vj reviewed (changes requested):
see inline comments

Commit:
promdump: update test suite to be more declarative

Updating the test suite to a more declarative style.

The list of tests to run, their arguments and expected returns are now built using a set of structs (though pointers need to be populated separately), then executed in a slightly cryptic foreach-style "range" loop.

This approach is cleaner but does come at a cost of reduced flexibility in case we have tests that need to be "different" for some reason.
